### PR TITLE
Make Reset a trait

### PR DESCRIPTION
**Pull request by jackkoenig**
Bool implements Reset. Compatibility package includes an implicit
conversion from Reset to Bool.

### CODE REVIEW

**Comment by jackkoenig** [comment]
Yeah I tried making ImplicitModule clone the type of the current reset, but that field is cleared by Module.apply so we'll have to think about how to do that properly.

I believe that it's generally okay to cast even AsyncReset to Bool as we do internally for printfs and asserts but yeah people need to be careful when doing so in general logic. I will make sure to look into this when PRing AsyncReset.

**Comment by ucbjrl** [comment]
This appears to break rocket-chip:
```shell
[info] Compiling 199 Scala sources to /vm/home/jenkins/workspace/rocket-chip_with_chisel3/target/scala-2.11/classes...
[error] /vm/home/jenkins/workspace/rocket-chip_with_chisel3/src/main/scala/jtag/JtagTap.scala:103: type mismatch;
[error]  found   : chisel3.core.Reset
[error]  required: chisel3.core.Bool
[error]   when (reset) {
[error]         ^
[error] one error found
[error] (rocketchip/compile:compileIncremental) Compilation failed
[error] Total time: 73 s, completed Aug 17, 2017 12:14:55 PM
make: *** [/vm/home/jenkins/workspace/rocket-chip_with_chisel3/vsim/generated-src/freechips.rocketchip.system.DefaultConfig.fir] Error 1
Build step 'Execute shell' marked build as failure
```

**Comment by ucbjrl** [comment]
This appears to break chisel-testers:
```shell
[info] Compiling 34 Scala sources to /vm/home/jenkins/sharedspace/chisel3-full/chisel-testers/target/scala-2.11/test-classes...
[error] /vm/home/jenkins/sharedspace/chisel3-full/chisel-testers/src/test/scala/examples/DynamicMemorySearch.scala:23: type mismatch;
[error]  found   : chisel3.core.Reset
[error]  required: chisel3.core.Bool
[error]   when(reset) {
[error]        ^
[error] /vm/home/jenkins/sharedspace/chisel3-full/chisel-testers/src/test/scala/examples/Router.scala:56: type mismatch;
[error]  found   : chisel3.core.Reset
[error]  required: chisel3.core.Bool
[error]   when(reset) {
[error]        ^
[error] two errors found
[error] (test:compileIncremental) Compilation failed
[error] Total time: 23 s, completed Aug 17, 2017 11:49:55 AM
Build step 'Execute shell' marked build as failure
```

**Comment by jackkoenig** [comment]
Ah yes, those are expected, need explicit `.toBool` casts. I was planning to PR a fix to rocket-chip soon (with accompanying bump), I'll PR a chisel-testers fix too

File: src/test/scala/chiselTests/CompatibilitySpec.scala [hunk 1]
```diff
@@ -260,4 +260,15 @@ class CompatibiltySpec extends ChiselFlatSpec with GeneratorDrivenPropertyChecks
       io.out := RegEnable(io.in(0), true.B)
     })
   }
+
+  "Reset" should "still walk, talk, and quack like a Bool" in {
```
Comment (edwardcwang) [inline src/test/scala/chiselTests/CompatibilitySpec.scala]: quack

**Comment by ucbjrl** [comment]
The casts work for reading values, but what about setting them?

See #709

**Pull request merged**